### ISSUE TITLE
PoolThreadCaches aren't getting freed?

### ISSUE THREAD

**Issue by yaud** [netty/netty#6819]
### Expected behavior
PoolThreadCache should be freed when thread it is bound to terminates.

### Actual behavior
We observed off-heap memory leak in our application that is run in Spark 1.6.2 on Amazon EMR 4.8.0.
Analyzing driver's heap dump I noticed that PoolThreadLocalCache's caches counter doesn't decrease which makes me thinking that PoolThreadCache is not getting freed:
![image](https://cloud.githubusercontent.com/assets/7288089/26827632/ce5bf81a-4ac6-11e7-8148-f25b0af0a2a5.png)
At the same time there are not that many threads:
![image](https://cloud.githubusercontent.com/assets/7288089/26827663/fb4b4eac-4ac6-11e7-9614-d581e023d450.png)

### Steps to reproduce
I don't think it is specific to our application but rather to Spark/Netty itself.

### Minimal yet complete reproducer code (or URL to code)
I can provide heap dumps if needed.

### Netty version
4.0.29.Final

### JVM version (e.g. `java -version`)
java version "1.8.0_45"
Java(TM) SE Runtime Environment (build 1.8.0_45-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.45-b02, mixed mode)

### OS version (e.g. `uname -a`)
Amazon (emr-4.8.0), Linux 4.4.11-23.53.amzn1.x86_64 #1 SMP Wed Jun 1 22:22:50 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by Scottmitch** [comment]
4.0.29.Final is relatively old ([2015-06-23 04:44](https://repo1.maven.org/maven2/io/netty/netty-all/4.0.29.Final/)). Can you update to the latest Netty version? Also you may want to experiment with [io.netty.allocator.useCacheForAllThreads](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L114) system property. This can limit the pool thread cache only be used for threads of type `FastThreadLocalThread`.

**Comment by yaud** [comment]
@Scottmitch unfortunately, I can not upgrade to a newer version of Netty as Spark on EMR comes bundled with it.
Also that property is not available in 4.0.29.Final.

**Comment by normanmaurer** [comment]
Then maybe you should create an issue in spark ?


On 6. June 2017 at 20:28:25, Yauheni Audzeichyk (notifications@github.com)
wrote:

@Scottmitch <https://github.com/scottmitch> unfortunately, I can not
upgrade to a newer version of Netty as Spark on EMR comes bundled with it.
Also that property is not available in 4.0.29.Final.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/netty/netty/issues/6819#issuecomment-306575848>, or mute
the thread

**Comment by yaud** [comment]
@normanmaurer that was my initial thought. However, after reviewing Spark's sources I didn't find anything suspicious. They just create an instance of PooledByteBufAllocator and use it to bootstrap Netty channel.
I understand that I should try upgrading either Netty or Spark but it is not feasible in my case. Spark itself gives a little of control on this piece - I tried setting spark.shuffle.io.preferDirectBufs=false which turns off usage of direct arenas in Netty but it didn't help.

I wonder if there are some Netty's properties that I could try out. Thanks.

**Comment by Scottmitch** [comment]
> However, after reviewing Spark's sources I didn't find anything suspicious.

The issue may not be in spark, but in the older Netty version it is using. Many things have changed since the Netty 4.0.29.Final. Please work with the spark folks to update if you want support from the Netty community.

**Comment by zsxwing** [comment]
I think this is probably #5833

By the way, Spark already updated to 4.0.43.Final on master branch.

**Comment by yaud** [comment]
@Scottmitch @zsxwing thanks! Fix for [#5833](https://github.com/netty/netty/issues/5833) is already included in Netty 4.0.42.Final which is used by Spark 2.1.0 so I ported our application to the latest Spark 2.1.1 and gave it a try. It's been running for 18 hours so far without any obvious off-heap leak.

Closing this thread as it looks like the latest Spark doesn't suffer from this issue anymore. Thanks for your help guys!
